src/message/getHtmlPieceDescriptors.js

```diff
@@ -7,17 +7,17 @@ import { isSameDay } from '../utils/date';
 export default (
   messages: $ReadOnlyArray<Message | Outbox>,
   narrow: Narrow,
-): HtmlPieceDescriptor[] => {
+): $ReadOnlyArray<HtmlPieceDescriptor> => {
   const showHeader = !isPmNarrow(narrow) && !isTopicNarrow(narrow);
 
-  let prevMessage = undefined;
-  const sections = [{ key: 0, data: [], message: null }];
-  messages.forEach(message => {
+  return messages.flatMap((message, index, arr) => {
```

Comment:
I'm happy to see `prevMessage` turned into a `const` at the top of the loop body, instead of a piece of state that gets mutated.

But the thing about this `flatMap` approach is that it'll create a lot of little arrays (mostly of length 1) which promptly get thrown away; one of them per message. That's quite a bit of allocation, and allocation (plus the GC work it makes necessary) is typically a large fraction of the CPU work an application ends up having to do. That means it's likely that the `flatMap` version of this function is a noticeable constant factor slower -- I'd guess it's somewhere between 10% slower and 2x slower -- than a version that pushes directly onto a single array.

Because this codepath where we're computing what to show in the message list is a potentially performance-sensitive area, it's one where it's worth taking care with that kind of constant factor as long as it doesn't make the source code substantially worse.

I'll push a quick WIP commit onto the branch which converts this to a plain loop but keeps the non-stateful `prevMessage`. Then please go ahead and squash that into the appropriate spot earlier in the branch, and/or make similar changes in a different order.